### PR TITLE
[improve][io] ElasticSearch sink: conditional id hashing

### PR DESCRIPTION
**Pull request by nicoloboschi**
### Motivation
#15428 introduced a new flag for hashing the document id. This is a great option for not getting into the hard limit of 512 bytes of ElasticSearch. 

In prod environment this flag is likely to not be enabled without troubles. If the sink was working before without that flag, all the doc ids are formatted as JSON format. With the same key, an update or deletion would be triggered on ES side. If the user enables the hashing option, those ids won't matching anymore and it will end up creating new records.

In some cases this is not a viable option. Since the hashing algo is actually useful only for workaround the ES limit, it makes sense to have an option to only hash if the would be key length errors on ES side. 
In this way, in a living cluster the hashing could be turned on with 100% compatibility with the existing documents.
- The <= 512 bytes record will behave the same way
- For > 512 bytes record, it's guaranteed the record is not present on elastic. Enabling this new option, the record will be inserted with a hashed (base64 format) key.

The collision between an hashed key and a non hashed key is not possible:
- hashed keys are represented in base64 (no { } characters allowed)
- non hashed keys are json     

To sum up, if a living env hits the issue, these are the suggested steps to get the sink to work again without losing messages and unblocking the subscription:
1. Stop the sink
2. Change the following config properties:
- `canonicalKeyFields=true`
- `idHashingAlgorithm=SHA256`
- `conditionalIdHashing=true`
Note that changing may alter the key format and therefore creates new records but it's required to get the hashing option to works correctly.
3. Restart the sink

### Modifications

* New option `conditionalIdHashing` default to false. If the raw key is <= 512 bytes and idHashingAlgorithm is set, then the hashing won't be performed  

### Documentation

- [x] `doc` 
- [ ] `doc-required` 
- [ ] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

File: pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchSink.java [hunk 1]
```diff
@@ -240,20 +240,28 @@ public Pair<String, String> extractIdAndDocument(Record<GenericObject> record) t
...
-                    case SHA512:
-                        hasher = Hashing.sha512().newHasher();
-                        break;
-                    default:
-                        throw new UnsupportedOperationException("Unsupported IdHashingAlgorithm: "
-                                + idHashingAlgorithm);
+
+                boolean performHashing = true;
+                if (elasticSearchConfig.isConditionalIdHashing()
+                        && id.getBytes(StandardCharsets.UTF_8).length <= 512) {
```
Comment (eolivelli) [inline pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchSink.java:246]: do we really need to create the byte[] instance ?
it will generate some garbage

maybe you can create the byte[] here and do not call  `hasher.putString(id, StandardCharsets.UTF_8);` but use the byte[] created here.
I suspect that `putString` will perform the encoding another time

File: site2/docs/io-elasticsearch-sink.md [hunk 2]
```diff
@@ -52,43 +52,44 @@ The configuration of the Elasticsearch sink connector has the following properti
 
 ### Property
 
-| Name | Type|Required | Default | Description
```
Comment (eolivelli) [inline site2/docs/io-elasticsearch-sink.md:55]: did you reformat the table ?

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchSink.java (hunk 1 above)
Comment (nicoloboschi) [reply pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchSink.java]: good catch!

File: site2/docs/io-elasticsearch-sink.md (hunk 2 above)
Comment (nicoloboschi) [reply site2/docs/io-elasticsearch-sink.md:55]: yes it was malformed, the IDE did it. I just rendered the table locally and it looks good

**Review by eolivelli — APPROVED** [review]
Lgtm

**Comment by lhotari** [comment]
@nicoloboschi @eolivelli I'm just wondering if the behavior of `conditionalIdHashing` should be revisited slightly. In most cases, the id will be the clear text one. However when the length goes over 512 bytes, it becomes a hash value. The length of the resulting base64 encoded value is fairly short. 

For example in the test, the `AkJcD1sNq/PSuRFfP3cjoCrYvPsVNKDSMWFP1CuBiPY` is an example and the length is 43 bytes. Why not take a prefix (512-43 bytes=469 bytes) from the original value and add the hash to the end to make it a unique 512 bytes value?
This probably doesn't matter that much, but it could make the id values more human readable.
